{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a03d7acf-f81b-4f0b-be46-22e0f0bc8c10",
   "metadata": {},
   "source": [
    "# Preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "92dbb37b-f897-4934-906e-969076240c84",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Installing package into ‘/home/strawberry/R/x86_64-pc-linux-gnu-library/4.4’\n",
      "(as ‘lib’ is unspecified)\n",
      "\n",
      "Installing package into ‘/home/strawberry/R/x86_64-pc-linux-gnu-library/4.4’\n",
      "(as ‘lib’ is unspecified)\n",
      "\n",
      "Installing package into ‘/home/strawberry/R/x86_64-pc-linux-gnu-library/4.4’\n",
      "(as ‘lib’ is unspecified)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Manually installing some packages to see if I can make it work\n",
    "install.packages(\"boot\")\n",
    "install.packages(\"cluster\")\n",
    "install.packages(\"codetools\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "14665da0-c0c9-410c-b3a7-d1af323be4ea",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Installing package into ‘/home/strawberry/R/x86_64-pc-linux-gnu-library/4.4’\n",
      "(as ‘lib’ is unspecified)\n",
      "\n",
      "Installing package into ‘/home/strawberry/R/x86_64-pc-linux-gnu-library/4.4’\n",
      "(as ‘lib’ is unspecified)\n",
      "\n",
      "Installing package into ‘/home/strawberry/R/x86_64-pc-linux-gnu-library/4.4’\n",
      "(as ‘lib’ is unspecified)\n",
      "\n",
      "Installing package into ‘/home/strawberry/R/x86_64-pc-linux-gnu-library/4.4’\n",
      "(as ‘lib’ is unspecified)\n",
      "\n",
      "Installing package into ‘/home/strawberry/R/x86_64-pc-linux-gnu-library/4.4’\n",
      "(as ‘lib’ is unspecified)\n",
      "\n",
      "Installing package into ‘/home/strawberry/R/x86_64-pc-linux-gnu-library/4.4’\n",
      "(as ‘lib’ is unspecified)\n",
      "\n",
      "Installing package into ‘/home/strawberry/R/x86_64-pc-linux-gnu-library/4.4’\n",
      "(as ‘lib’ is unspecified)\n",
      "\n",
      "Installing package into ‘/home/strawberry/R/x86_64-pc-linux-gnu-library/4.4’\n",
      "(as ‘lib’ is unspecified)\n",
      "\n",
      "Installing package into ‘/home/strawberry/R/x86_64-pc-linux-gnu-library/4.4’\n",
      "(as ‘lib’ is unspecified)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "install.packages(\"foreign\")\n",
    "install.packages(\"KernSmooth\")\n",
    "install.packages(\"lattice\")\n",
    "install.packages(\"MASS\")\n",
    "install.packages(\"Matrix\")\n",
    "install.packages(\"mgcv\")\n",
    "install.packages(\"nlme\")\n",
    "install.packages(\"rpart\")\n",
    "install.packages(\"survival\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "5063137c-6608-4903-8119-5fb6e55ce52d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Installing package into ‘/home/strawberry/R/x86_64-pc-linux-gnu-library/4.4’\n",
      "(as ‘lib’ is unspecified)\n",
      "\n",
      "also installing the dependencies ‘htmlwidgets’, ‘mockery’, ‘whoami’\n",
      "\n",
      "\n",
      "Installing package into ‘/home/strawberry/R/x86_64-pc-linux-gnu-library/4.4’\n",
      "(as ‘lib’ is unspecified)\n",
      "\n",
      "Installing package into ‘/home/strawberry/R/x86_64-pc-linux-gnu-library/4.4’\n",
      "(as ‘lib’ is unspecified)\n",
      "\n",
      "also installing the dependencies ‘rjson’, ‘tinytest’, ‘simplermarkdown’\n",
      "\n",
      "\n",
      "Installing package into ‘/home/strawberry/R/x86_64-pc-linux-gnu-library/4.4’\n",
      "(as ‘lib’ is unspecified)\n",
      "\n",
      "Installing package into ‘/home/strawberry/R/x86_64-pc-linux-gnu-library/4.4’\n",
      "(as ‘lib’ is unspecified)\n",
      "\n",
      "also installing the dependency ‘unitizer’\n",
      "\n",
      "\n",
      "Installing package into ‘/home/strawberry/R/x86_64-pc-linux-gnu-library/4.4’\n",
      "(as ‘lib’ is unspecified)\n",
      "\n",
      "also installing the dependencies ‘plogr’, ‘RSQLite’\n",
      "\n",
      "\n",
      "Warning message in init_backup_env():\n",
      "“init_backup_env called a second time after init_shadowenv:\n",
      "44: fun(libname, pkgname)\n",
      "43: doTryCatch(return(expr), name, parentenv, handler)\n",
      "42: tryCatchOne(expr, names, parentenv, handlers[[1L]])\n",
      "41: tryCatchList(expr, classes, parentenv, handlers)\n",
      "40: tryCatch(fun(libname, pkgname), error = identity)\n",
      "39: runHook(\".onLoad\", env, package.lib, package)\n",
      "38: loadNamespace(package, lib.loc)\n",
      "37: doTryCatch(return(expr), name, parentenv, handler)\n",
      "36: tryCatchOne(expr, names, parentenv, handlers[[1L]])\n",
      "35: tryCatchList(expr, classes, parentenv, handlers)\n",
      "34: tryCatch({\n",
      "        attr(package, \"LibPath\") <- which.lib.loc\n",
      "        ns <- loadNamespace(package, lib.loc)\n",
      "        env <- attachNamespace(ns, pos = pos, deps, exclude, include.only)\n",
      "    }, error = function(e) {\n",
      "        P <- if (!is.null(cc <- conditionCall(e))) \n",
      "            paste(\" in\", deparse(cc)[1L])\n",
      "        else \"\"\n",
      "        msg <- gettextf(\"package or namespace load failed for %s%s:\\n %s\", \n",
      "            sQuote(package), P, conditionMessage(e))\n",
      "        if (logical.return && !quietly) \n",
      "            message(paste(\"Error:\", msg), domain = NA)\n",
      "        else stop(msg, call. = FALSE, domain = NA)\n",
      "    })\n",
      "33: library(package, lib.loc = lib.loc, character.only = TRUE, logical.return = TRUE, \n",
      "        warn.conflicts = warn.conflicts, quietly = quietly, mask.ok = mask.ok, \n",
      "        exclude = exclude, include.only = include.only, attach.required = attach.required)\n",
      "32: doTryCatch(return(expr), name, parentenv, handler)\n",
      "31: tryCatchOne(expr, names, parentenv, handlers[[1L]])\n",
      "30: tryCatchList(expr, classes, parentenv, handlers)\n",
      "29: tryCatch(library(package, lib.loc = lib.loc, character.only = TRUE, \n",
      "        logical.return = TRUE, warn.conflicts = warn.conflicts, quietly = quietly, \n",
      "        mask.ok = mask.ok, exclude = exclude, include.only = include.only, \n",
      "        attach.required = attach.required), error = function(e) e)\n",
      "28: require(pkg, character.only = TRUE, quietly = TRUE) at <text>#13\n",
      "27: FUN(X[[i]], ...)\n",
      "26: lapply(packages, install_packages)\n",
      "25: eval(expr, envir, enclos)\n",
      "24: eval(expr, envir, enclos)\n",
      "23: eval_with_user_handlers(expr, envir, enclos, user_handlers)\n",
      "22: withVisible(eval_with_user_handlers(expr, envir, enclos, user_handlers))\n",
      "21: withCallingHandlers(withVisible(eval_with_user_handlers(expr, \n",
      "        envir, enclos, user_handlers)), warning = wHandler, error = eHandler, \n",
      "        message = mHandler)\n",
      "20: doTryCatch(return(expr), name, parentenv, handler)\n",
      "19: tryCatchOne(expr, names, parentenv, handlers[[1L]])\n",
      "18: tryCatchList(expr, classes, parentenv, handlers)\n",
      "17: tryCatch(expr, error = function(e) {\n",
      "        call <- conditionCall(e)\n",
      "        if (!is.null(call)) {\n",
      "            if (identical(call[[1L]], quote(doTryCatch))) \n",
      "                call <- sys.call(-4L)\n",
      "            dcall <- deparse(call, nlines = 1L)\n",
      "            prefix <- paste(\"Error in\", dcall, \": \")\n",
      "            LONG <- 75L\n",
      "            sm <- strsplit(conditionMessage(e), \"\\n\")[[1L]]\n",
      "            w <- 14L + nchar(dcall, type = \"w\") + nchar(sm[1L], type = \"w\")\n",
      "            if (is.na(w)) \n",
      "                w <- 14L + nchar(dcall, type = \"b\") + nchar(sm[1L], \n",
      "                    type = \"b\")\n",
      "            if (w > LONG) \n",
      "                prefix <- paste0(prefix, \"\\n  \")\n",
      "        }\n",
      "        else prefix <- \"Error : \"\n",
      "        msg <- paste0(prefix, conditionMessage(e), \"\\n\")\n",
      "        .Internal(seterrmessage(msg[1L]))\n",
      "        if (!silent && isTRUE(getOption(\"show.error.messages\"))) {\n",
      "            cat(msg, file = outFile)\n",
      "            .Internal(printDeferredWarnings())\n",
      "        }\n",
      "        invisible(structure(msg, class = \"try-error\", condition = e))\n",
      "    })\n",
      "16: try(f, silent = TRUE)\n",
      "15: handle(ev <- withCallingHandlers(withVisible(eval_with_user_handlers(expr, \n",
      "        envir, enclos, user_handlers)), warning = wHandler, error = eHandler, \n",
      "        message = mHandler))\n",
      "14: timing_fn(handle(ev <- withCallingHandlers(withVisible(eval_with_user_handlers(expr, \n",
      "        envir, enclos, user_handlers)), warning = wHandler, error = eHandler, \n",
      "        message = mHandler)))\n",
      "13: evaluate_call(expr, parsed$src[[i]], envir = envir, enclos = enclos, \n",
      "        debug = debug, last = i == length(out), use_try = stop_on_error != \n",
      "            2L, keep_warning = keep_warning, keep_message = keep_message, \n",
      "        log_echo = log_echo, log_warning = log_warning, output_handler = output_handler, \n",
      "        include_timing = include_timing)\n",
      "12: evaluate(request$content$code, envir = .GlobalEnv, output_handler = oh, \n",
      "        stop_on_error = 1L)\n",
      "11: doTryCatch(return(expr), name, parentenv, handler)\n",
      "10: tryCatchOne(expr, names, parentenv, handlers[[1L]])\n",
      "9: tryCatchList(expr, names[-nh], parentenv, handlers[-nh])\n",
      "8: doTryCatch(return(expr), name, parentenv, handler)\n",
      "7: tryCatchOne(tryCatchList(expr, names[-nh], parentenv, handlers[-nh]), \n",
      "       names[nh], parentenv, handlers[[nh]])\n",
      "6: tryCatchList(expr, classes, parentenv, handlers)\n",
      "5: tryCatch(evaluate(request$content$code, envir = .GlobalEnv, output_handler = oh, \n",
      "       stop_on_error = 1L), interrupt = function(cond) {\n",
      "       log_debug(\"Interrupt during execution\")\n",
      "       interrupted <<- TRUE\n",
      "   }, error = .self$handle_error)\n",
      "4: executor$execute(msg)\n",
      "3: handle_shell()\n",
      "2: kernel$run()\n",
      "1: IRkernel::main()”\n",
      "Installing package into ‘/home/strawberry/R/x86_64-pc-linux-gnu-library/4.4’\n",
      "(as ‘lib’ is unspecified)\n",
      "\n",
      "also installing the dependencies ‘proxy’, ‘e1071’, ‘wk’, ‘R.methodsS3’, ‘R.oo’, ‘R.utils’, ‘R.cache’, ‘classInt’, ‘s2’, ‘units’, ‘R.rsp’, ‘sf’\n",
      "\n",
      "\n",
      "Warning message in install.packages(pkg, dependencies = TRUE):\n",
      "“installation of package ‘units’ had non-zero exit status”\n",
      "Warning message in install.packages(pkg, dependencies = TRUE):\n",
      "“installation of package ‘sf’ had non-zero exit status”\n",
      "This is loo version 2.8.0\n",
      "\n",
      "- Online documentation and vignettes at mc-stan.org/loo\n",
      "\n",
      "- As of v2.0.0 loo defaults to 1 core but we recommend using as many as possible. Use the 'cores' argument or set options(mc.cores = NUM_CORES) for an entire session. \n",
      "\n",
      "\n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:checkmate’:\n",
      "\n",
      "    anyMissing\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘munsell’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:colorspace’:\n",
      "\n",
      "    desaturate\n",
      "\n",
      "\n",
      "Installing package into ‘/home/strawberry/R/x86_64-pc-linux-gnu-library/4.4’\n",
      "(as ‘lib’ is unspecified)\n",
      "\n",
      "\n",
      "Attaching package: ‘RcppParallel’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Rcpp’:\n",
      "\n",
      "    LdFlags\n",
      "\n",
      "\n",
      "Installing package into ‘/home/strawberry/R/x86_64-pc-linux-gnu-library/4.4’\n",
      "(as ‘lib’ is unspecified)\n",
      "\n",
      "also installing the dependencies ‘terra’, ‘crosstalk’, ‘leaflet.providers’, ‘raster’, ‘sp’, ‘Cairo’, ‘leaflet’\n",
      "\n",
      "\n",
      "Warning message in install.packages(pkg, dependencies = TRUE):\n",
      "“installation of package ‘terra’ had non-zero exit status”\n",
      "Warning message in install.packages(pkg, dependencies = TRUE):\n",
      "“installation of package ‘Cairo’ had non-zero exit status”\n",
      "Warning message in install.packages(pkg, dependencies = TRUE):\n",
      "“installation of package ‘raster’ had non-zero exit status”\n",
      "Warning message in install.packages(pkg, dependencies = TRUE):\n",
      "“installation of package ‘leaflet’ had non-zero exit status”\n",
      "Installing package into ‘/home/strawberry/R/x86_64-pc-linux-gnu-library/4.4’\n",
      "(as ‘lib’ is unspecified)\n",
      "\n",
      "also installing the dependencies ‘credentials’, ‘zip’, ‘gitcreds’, ‘httr2’, ‘ini’, ‘gert’, ‘gh’, ‘whisker’, ‘usethis’\n",
      "\n",
      "\n",
      "\n",
      "rstan version 2.32.6 (Stan version 2.32.2)\n",
      "\n",
      "\n",
      "For execution on a local, multicore CPU with excess RAM we recommend calling\n",
      "options(mc.cores = parallel::detectCores()).\n",
      "To avoid recompilation of unchanged Stan programs, we recommend calling\n",
      "rstan_options(auto_write = TRUE)\n",
      "For within-chain threading using `reduce_sum()` or `map_rect()` Stan functions,\n",
      "change `threads_per_chain` option:\n",
      "rstan_options(threads_per_chain = 1)\n",
      "\n",
      "\n",
      "Installing package into ‘/home/strawberry/R/x86_64-pc-linux-gnu-library/4.4’\n",
      "(as ‘lib’ is unspecified)\n",
      "\n",
      "Installing package into ‘/home/strawberry/R/x86_64-pc-linux-gnu-library/4.4’\n",
      "(as ‘lib’ is unspecified)\n",
      "\n",
      "also installing the dependencies ‘downlit’, ‘pkgdown’, ‘zeallot’\n",
      "\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<ol>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate}\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\end{enumerate}\n"
      ],
      "text/markdown": [
       "1. NULL\n",
       "2. NULL\n",
       "3. NULL\n",
       "4. NULL\n",
       "5. NULL\n",
       "6. NULL\n",
       "7. NULL\n",
       "8. NULL\n",
       "9. NULL\n",
       "10. NULL\n",
       "11. NULL\n",
       "12. NULL\n",
       "13. NULL\n",
       "14. NULL\n",
       "15. NULL\n",
       "16. NULL\n",
       "17. NULL\n",
       "18. NULL\n",
       "19. NULL\n",
       "20. NULL\n",
       "21. NULL\n",
       "22. NULL\n",
       "23. NULL\n",
       "24. NULL\n",
       "25. NULL\n",
       "26. NULL\n",
       "27. NULL\n",
       "28. NULL\n",
       "29. NULL\n",
       "30. NULL\n",
       "31. NULL\n",
       "32. NULL\n",
       "33. NULL\n",
       "34. NULL\n",
       "35. NULL\n",
       "36. NULL\n",
       "37. NULL\n",
       "38. NULL\n",
       "39. NULL\n",
       "40. NULL\n",
       "41. NULL\n",
       "42. NULL\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[[1]]\n",
       "NULL\n",
       "\n",
       "[[2]]\n",
       "NULL\n",
       "\n",
       "[[3]]\n",
       "NULL\n",
       "\n",
       "[[4]]\n",
       "NULL\n",
       "\n",
       "[[5]]\n",
       "NULL\n",
       "\n",
       "[[6]]\n",
       "NULL\n",
       "\n",
       "[[7]]\n",
       "NULL\n",
       "\n",
       "[[8]]\n",
       "NULL\n",
       "\n",
       "[[9]]\n",
       "NULL\n",
       "\n",
       "[[10]]\n",
       "NULL\n",
       "\n",
       "[[11]]\n",
       "NULL\n",
       "\n",
       "[[12]]\n",
       "NULL\n",
       "\n",
       "[[13]]\n",
       "NULL\n",
       "\n",
       "[[14]]\n",
       "NULL\n",
       "\n",
       "[[15]]\n",
       "NULL\n",
       "\n",
       "[[16]]\n",
       "NULL\n",
       "\n",
       "[[17]]\n",
       "NULL\n",
       "\n",
       "[[18]]\n",
       "NULL\n",
       "\n",
       "[[19]]\n",
       "NULL\n",
       "\n",
       "[[20]]\n",
       "NULL\n",
       "\n",
       "[[21]]\n",
       "NULL\n",
       "\n",
       "[[22]]\n",
       "NULL\n",
       "\n",
       "[[23]]\n",
       "NULL\n",
       "\n",
       "[[24]]\n",
       "NULL\n",
       "\n",
       "[[25]]\n",
       "NULL\n",
       "\n",
       "[[26]]\n",
       "NULL\n",
       "\n",
       "[[27]]\n",
       "NULL\n",
       "\n",
       "[[28]]\n",
       "NULL\n",
       "\n",
       "[[29]]\n",
       "NULL\n",
       "\n",
       "[[30]]\n",
       "NULL\n",
       "\n",
       "[[31]]\n",
       "NULL\n",
       "\n",
       "[[32]]\n",
       "NULL\n",
       "\n",
       "[[33]]\n",
       "NULL\n",
       "\n",
       "[[34]]\n",
       "NULL\n",
       "\n",
       "[[35]]\n",
       "NULL\n",
       "\n",
       "[[36]]\n",
       "NULL\n",
       "\n",
       "[[37]]\n",
       "NULL\n",
       "\n",
       "[[38]]\n",
       "NULL\n",
       "\n",
       "[[39]]\n",
       "NULL\n",
       "\n",
       "[[40]]\n",
       "NULL\n",
       "\n",
       "[[41]]\n",
       "NULL\n",
       "\n",
       "[[42]]\n",
       "NULL\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# List of packages to install\n",
    "packages <- c(\n",
    "  \"backports\", \"BH\", \"callr\", \"checkmate\", \"cli\", \"colorspace\", \"cpp11\", \n",
    "  \"crayon\", \"desc\", \"digest\", \"evaluate\", \"fansi\", \"farver\", \"fastmap\", \n",
    "  \"ggplot2\", \"glue\", \"gtable\", \"htmltools\", \"IRkernel\", \"isoband\", \"jsonlite\", \n",
    "  \"loo\", \"matrixStats\", \"munsell\", \"pbdZMQ\", \"pkgbuild\", \"processx\", \"ps\", \n",
    "  \"QuickJSR\", \"Rcpp\", \"RcppEigen\", \"RcppParallel\", \"repr\", \"rlang\", \"rstan\", \n",
    "  \"scales\", \"StanHeaders\", \"svglite\", \"systemfonts\", \"uuid\", \"vctrs\", \"withr\"\n",
    ")\n",
    "\n",
    "# Install the packages\n",
    "install_packages <- function(pkg) {\n",
    "  if (!require(pkg, character.only = TRUE, quietly = TRUE)) {\n",
    "    install.packages(pkg, dependencies = TRUE)\n",
    "  }\n",
    "}\n",
    "\n",
    "# Install all packages\n",
    "lapply(packages, install_packages)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "484fb97d-c49e-46bd-a6c1-41998f543e81",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
      "Replacement repositories:\n",
      "    CRAN: https://cloud.r-project.org\n",
      "\n",
      "Bioconductor version 3.19 (BiocManager 1.30.24), R 4.4.1 (2024-06-14)\n",
      "\n",
      "Installing package(s) 'ShortRead'\n",
      "\n",
      "also installing the dependencies ‘Rsamtools’, ‘GenomicAlignments’, ‘Rhtslib’\n",
      "\n",
      "\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘Rhtslib’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘Rsamtools’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘GenomicAlignments’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘ShortRead’ had non-zero exit status”\n",
      "Installation paths not writeable, unable to update packages\n",
      "  path: /usr/lib/R/library\n",
      "  packages:\n",
      "    boot, cluster, codetools, foreign, KernSmooth, lattice, MASS, Matrix, mgcv,\n",
      "    nlme, rpart, survival\n",
      "  path: /usr/lib/R/site-library\n",
      "  packages:\n",
      "    backports, BH, callr, checkmate, cli, colorspace, cpp11, crayon, desc,\n",
      "    digest, evaluate, fansi, farver, fastmap, ggplot2, glue, gtable, htmltools,\n",
      "    IRkernel, isoband, jsonlite, loo, matrixStats, munsell, pbdZMQ, pkgbuild,\n",
      "    processx, ps, QuickJSR, Rcpp, RcppEigen, RcppParallel, repr, rlang, rstan,\n",
      "    scales, StanHeaders, svglite, systemfonts, uuid, vctrs, withr\n",
      "\n",
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
      "Replacement repositories:\n",
      "    CRAN: https://cloud.r-project.org\n",
      "\n",
      "Bioconductor version 3.19 (BiocManager 1.30.24), R 4.4.1 (2024-06-14)\n",
      "\n",
      "Installing package(s) 'dada2'\n",
      "\n",
      "also installing the dependencies ‘Rsamtools’, ‘GenomicAlignments’, ‘Rhtslib’, ‘ShortRead’\n",
      "\n",
      "\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘Rhtslib’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘Rsamtools’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘GenomicAlignments’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘ShortRead’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘dada2’ had non-zero exit status”\n",
      "Installation paths not writeable, unable to update packages\n",
      "  path: /usr/lib/R/library\n",
      "  packages:\n",
      "    boot, cluster, codetools, foreign, KernSmooth, lattice, MASS, Matrix, mgcv,\n",
      "    nlme, rpart, survival\n",
      "  path: /usr/lib/R/site-library\n",
      "  packages:\n",
      "    backports, BH, callr, checkmate, cli, colorspace, cpp11, crayon, desc,\n",
      "    digest, evaluate, fansi, farver, fastmap, ggplot2, glue, gtable, htmltools,\n",
      "    IRkernel, isoband, jsonlite, loo, matrixStats, munsell, pbdZMQ, pkgbuild,\n",
      "    processx, ps, QuickJSR, Rcpp, RcppEigen, RcppParallel, repr, rlang, rstan,\n",
      "    scales, StanHeaders, svglite, systemfonts, uuid, vctrs, withr\n",
      "\n"
     ]
    },
    {
     "ename": "ERROR",
     "evalue": "Error in library(ShortRead): there is no package called ‘ShortRead’\n",
     "output_type": "error",
     "traceback": [
      "Error in library(ShortRead): there is no package called ‘ShortRead’\nTraceback:\n",
      "1. library(ShortRead)"
     ]
    }
   ],
   "source": [
    "## Installing required R packages and activating them\n",
    "install.packages(\"tidyverse\")\n",
    "\n",
    "if (!require(\"BiocManager\", quietly = TRUE))\n",
    "    install.packages(\"BiocManager\")\n",
    "BiocManager::install(\"ShortRead\")\n",
    "\n",
    "if (!require(\"BiocManager\", quietly = TRUE))\n",
    "    install.packages(\"BiocManager\")\n",
    "BiocManager::install(\"dada2\")\n",
    "\n",
    "library(ShortRead)\n",
    "library(dada2)\n",
    "library(tidyverse)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1bcbbeeb-b114-4f5c-a52d-c31d82c286e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Defining parameters for trimminging and filtering\n",
    "minLength <- 1200 #Removes reads shorter than this length. Minimum length is enforced AFTER trimming\n",
    "maxLength <- 1800 #Removes reads longer than this length. Maximum length is enforced BEFORE trimming \n",
    "trimLeft <- 100 #The number of nucleotides to remove from the start of each read. Must cover the primer length\n",
    "trimRight <- 100 #The number of nucleotides to remove from the end of each read. Must cover the primer length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6f6189f-8865-46ba-935f-4f49e71abf63",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Handling the files\n",
    "path <- getwd()\n",
    "files <- list.files(pattern = \"Sample\") # create object for files based on patter \n",
    "sample.names <- tools::file_path_sans_ext(basename(files))  # remove extension\n",
    "rawFiles <- sort(list.files(path, pattern=\".fastq\", full.names = TRUE))\n",
    "filtFiles <- file.path(path, \"filtered\", paste0(sample.names, \"_filt.fastq\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c45ec966-aac5-4906-a02d-b2e16df523f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Sending filtered files to output directory\n",
    "### I previously created the directory called 'filtered' and saved the table \n",
    "out <- filterAndTrim(rawFiles, filtFiles, trimLeft = trimLeft, \n",
    "                     trimRight = trimRight, maxLen = maxLength, \n",
    "                     minLen = minLength,  truncQ = 0, \n",
    "                     compress = FALSE)\n",
    "out\n",
    "write.table(out, file = \"read_count_QC.txt\", sep = \"\\t\", row.names = TRUE)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
