# Microbiota analyses 
- Code adapted from Zaramela and Zorz (https://github.com/Microbial-Ecosystems-Lab/asthma_obesity and https://github.com/jkzorz/SituSeq/)

Open tidyverse environment (using conda), then open R in terminal --> install required packages; when finished, open Rstudio, also in terminal; This workflow usually works better than simply oppening Rstudio 

## Go to script 1. Preprocessing 

## Go to script 2. Database creation and Reads Alignment

## Go to script 3. Assigning taxonomy to nanopore reads
## Go to script 3.1 Calculate annotation frequency  

## Go to script 4. Diversity analysis

## Go to script 5. Visualize taxonomy 

## Go to script 7. Correlation heatmap 

Some mangrove isolates were selected for whole-genome sequencing (Illumina NovaSeq 6000). The assembly, annotation and all including analyses can be assessed in the following script.

## Go to script 8. WGS_processing
